examples/selective2d/2d_train.py

```diff
@@ -0,0 +1,354 @@
+  parser.add_argument('--pp_size', type=int, default=2)
+
+  parser.add_argument('--debug', dest='debug', action='store_true')
+
+  args = parser.parse_args()
+
+  return args
+
+
+def rank_print(x):
+    _rank = os.getenv('RANK')
+    if _rank == '0':
+        print(x)
+
+def get_rand(args):
+    x = torch.randint(0, args.vocab_size, (args.batch_size, args.block_size), device=args.device)
+    y = torch.randint(0, args.vocab_size, (args.batch_size, args.block_size), device=args.device)
+    return x, y
+
+def tp_attention(model, name, mesh, tp_dim=0, q='q', k='k', v='v', o='c_proj'):
+  layer = model.get_submodule(name)
+  parallelize_module(layer, mesh, { q: ColwiseParallel(),
+                                    k: ColwiseParallel(),
+                                    v: ColwiseParallel(),
+                                    o: RowwiseParallel() },
+                                    tp_mesh_dim=tp_dim)
+
+  return model
+
+def tp_mlp(model, name, mesh, tp_dim=0, mlp='mlp'):
+  layer = model.get_submodule(name)
+  parallelize_module(layer, mesh, { mlp: PairwiseParallel() }, tp_mesh_dim=tp_dim)
+
+  return model
+
+
+def tp(model, n_layer, mesh, offset=0, tp_dim=0):
+  for i in range(n_layer):
+    block = model.get_submodule(f'transformer.h.{i + offset}')
+    parallelize_module(block, mesh, {'attn.q': ColwiseParallel(),
+                                     'attn.k': ColwiseParallel(),
+                                     'attn.v': ColwiseParallel(),
+                                     'attn.c_proj': RowwiseParallel(),
+                                     'mlp': PairwiseParallel()},
+                                     tp_mesh_dim=tp_dim) 
+
+  return model
+
+def pp(model, pp_device_mesh, args):
+  from pippy.IR import annotate_split_points, PipeSplitWrapper 
+  from pippy import split_into_equal_size
+  from pippy.compile import compile_stage
+  from pippy.microbatch import TensorChunkSpec, sum_reducer
+
+  pp_chunks = args.world_size 
+  pp_groups = pp_device_mesh.get_dim_groups()[0]
+
+  output_chunk_spec = (TensorChunkSpec(0), sum_reducer)
+  stage = compile_stage(model, args.rank, args.world_size, pp_chunks, pp_device_mesh, pp_groups, 
+            example_inputs=[X, Y], 
+            output_chunk_spec=output_chunk_spec)
+  
+  print(f'[Rank{_rank}] {stage.submod.print_readable()}')
+  return model, stage 
+
+def pp_and_tp(model, mesh, args):
+  from pippy.compile import compile_stage
+  from pippy.microbatch import TensorChunkSpec, sum_reducer
+
+  pp_dim, tp_dim = 0, 1
+  pp_rank, tp_rank = args.rank // args.tp_size, args.rank % args.tp_size
+  pp_groups = mesh.get_dim_groups()[pp_dim]
+
+  # TP
+  tp(model, args.n_layer, mesh, 0, tp_dim)
+
+  X, Y = get_rand(args)
+
+  # PP
+  stage = compile_stage(model, pp_rank, args.world_size, args.pp_size, args.device, pp_groups, 
+            example_inputs=[X, Y],
+            )
+  
+  return model, stage
+
+def even_cut(model, args, pp_size, cut={}):
```

Comment:
Why `cut` is passed as an argument? And using `{}` as the default value is never good.

Reply:
I was thinking mixing two algorithms but at this point we may not need it. removed :)